cocotb/run.py

```diff
@@ -0,0 +1,175 @@
+
+
+def _build_lib(lib, dist):
+    dist.ext_modules = [lib]
+    _build_ext = build_ext(dist)
+    _build_ext.finalize_options()
+
+    _build_ext.run()
+    out_lib = _build_ext.get_outputs()
+
+    lib_name = lib.name
+    lib_path = os.path.abspath(out_lib[0])
+    dir_name = os.path.dirname(lib_path)
+    ext_name = os.path.splitext(lib_path)[1][1:]
+
+    print("Building:", out_lib[0])
+
+    _symlink_force(
+        lib_path, os.path.join(os.path.abspath(dir_name), lib_name + "." + ext_name)
+    )
+
+    return dir_name
+
+
+def build_libs():
+    share_dir = os.path.join(os.path.dirname(cocotb.__file__), "share")
+
+    python_lib = sysconfig.get_config_var("LDLIBRARY")
+    python_lib_link = os.path.splitext(python_lib)[0][3:]
+
+    dist = Distribution()
+    dist.parse_config_files()
+
+    libcocotbutils = Extension(
+        "libcocotbutils",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/utils/cocotb_utils.c"],
+    )
+
+    lib_path = _build_lib(libcocotbutils, dist)
+
+    libgpilog = Extension(
+        "libgpilog",
+        include_dirs=[share_dir + "/include"],
+        libraries=[python_lib_link, "pthread", "dl", "util", "rt", "m", "cocotbutils"],
+        library_dirs=[lib_path],
+        sources=[share_dir + "/lib/gpi_log/gpi_logging.c"],
+    )
+
+    _build_lib(libgpilog, dist)
+
+    libcocotb = Extension(
+        "libcocotb",
+        define_macros=[("PYTHON_SO_LIB", python_lib)],
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/embed/gpi_embed.c"],
+    )
+
+    _build_lib(libcocotb, dist)
+
+    libgpi = Extension(
+        "libgpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["cocotbutils", "gpilog", "cocotb", "stdc++"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/gpi/GpiCbHdl.cpp",
+            share_dir + "/lib/gpi/GpiCommon.cpp",
+        ],
+    )
+
+    _build_lib(libgpi, dist)
+
+    libsim = Extension(
+        "libsim",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/simulator/simulatormodule.c"],
+    )
+
+    _build_lib(libsim, dist)
+
+    libvpi = Extension(
+        "libvpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["gpi", "gpilog"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/vpi/VpiImpl.cpp",
+            share_dir + "/lib/vpi/VpiCbHdl.cpp",
+        ],
+    )
+
+    _build_lib(libvpi, dist)
+
+    _symlink_force(
+        os.path.join(os.path.abspath(lib_path), "libvpi.so"),
+        os.path.join(os.path.abspath(lib_path), "gpivpi.vpl"),
+    )
+    _symlink_force(
+        os.path.join(os.path.abspath(lib_path), "libvpi.so"),
+        os.path.join(os.path.abspath(lib_path), "cocotb.vpl"),
+    )
+    _symlink_force(
+        os.path.join(os.path.abspath(lib_path), "libsim.so"),
+        os.path.join(os.path.abspath(lib_path), "simulator.so"),
+    )
```

Comment:
Don't think will work on windows, sadly

Reply:
Right. There will be more things not working on Windows ;-(